mlflow/onnx.py

```diff
@@ -0,0 +1,194 @@
+
+FLAVOR_NAME = "onnx"
+
+DEFAULT_CONDA_ENV = _mlflow_conda_env(
+    additional_conda_deps=None,
+    additional_pip_deps=[
+        "onnx=={}".format(onnx.__version__),
+        # The ONNX pyfunc representation requires the OnnxRuntime
+        # inference engine. Therefore, the conda environment must
+        # include OnnxRuntime
+        "onnxruntime=={}".format(onnxrt.__version__),
+    ],
+    additional_conda_channels=None,
+)
+
+
+def save_model(onnx_model, path, conda_env=None, mlflow_model=Model()):
+    """
+    Save an ONNX model to a path on the local file system.
+
+    :param onnx_model: ONNX model to be saved.
+    :param path: Local path where the model is to be saved.
+    :param conda_env: Either a dictionary representation of a Conda environment or the path to a
+                      Conda environment yaml file. If provided, this decribes the environment
+                      this model should be run in. At minimum, it should specify the dependencies
+                      contained in ``mlflow.onnx.DEFAULT_CONDA_ENV``. If `None`, the default
+                      ``mlflow.onnx.DEFAULT_CONDA_ENV`` environment will be added to the model.
+                      The following is an *example* dictionary representation of a Conda
+                      environment::
+
+                        {
+                            'name': 'mlflow-env',
+                            'channels': ['defaults'],
+                            'dependencies': [
+                                'python=3.6.0',
+                                'onnx=1.4.1',
+                                'onnxruntime=0.3.0'
+                            ]
+                        }
+
+    :param mlflow_model: :py:mod:`mlflow.models.Model` this flavor is being added to.
+    """
+    path = os.path.abspath(path)
+    if os.path.exists(path):
+        raise Exception("Path '{}' already exists".format(path))
+    os.makedirs(path)
+    model_data_subpath = "model.onnx"
+    model_data_path = os.path.join(path, model_data_subpath)
+
+    # Save onnx-model
+    onnx.save_model(onnx_model, model_data_path)
+
+    conda_env_subpath = "conda.yaml"
+    if conda_env is None:
+        conda_env = DEFAULT_CONDA_ENV
+    elif not isinstance(conda_env, dict):
+        with open(conda_env, "r") as f:
+            conda_env = yaml.safe_load(f)
+    with open(os.path.join(path, conda_env_subpath), "w") as f:
+        yaml.safe_dump(conda_env, stream=f, default_flow_style=False)
+
+    pyfunc.add_to_model(mlflow_model, loader_module="mlflow.onnx",
+                        data=model_data_subpath, env=conda_env_subpath)
+    mlflow_model.add_flavor(FLAVOR_NAME, onnx_version=onnx.__version__, data=model_data_subpath)
+    mlflow_model.save(os.path.join(path, "MLmodel"))
+
+
+def _load_model(model_file):
+    onnx_model = onnx.load(model_file)
+    # Check Formation
+    onnx.checker.check_model(onnx_model)
+    return onnx_model
+
+
+class _OnnxModelWrapper:
```

Comment:
@avflor It occurred to me that we can get the input type for each tensor in the ONNX model via the  `onnxrt.InferenceSession` object. Then, we can check for tensors of type `tensor(float)` and cast the corresponding Pandas DataFrame contents accordingly (either a particular column for the multi-input case or the entire set of values for the single-input case).

Here's some code indicating that we can examine the `InferenceSession` object to obtain these types:
```
>>> rt = onnxrt.InferenceSession("my_model.onnx")
>>> rt.get_inputs()[0].type
'tensor(float)'
```

I can make this change if you'd like. Does this make sense from your perspective?

Reply:
@dbczumar This is great! Makes the code much cleaner. I quickly made the change and testing it now. I'll let you know if it works and then you can take another pass. Thanks a lot for your help!

Reply:
@dbczumar ONNX tests are passing: https://travis-ci.org/mlflow/mlflow/builds/538360806
Now running full tests in travis. Please take a pass and let me know how the code looks like. Also feel free to make any changes :)